### PR TITLE
[CALCITE-4092] NPE using WITH clause without a corresponding SELECT FROM (James Kim)

### PR DESCRIPTION
**Pull request by James-Jeyun-Kim**

### CODE REVIEW

**Comment by James-Jeyun-Kim** [comment]
This PR catches an invalid WITH clause during the query validation step so that it doesn't lead to a NPE later on in the sqlToRelConverter stage.

**Comment by James-Jeyun-Kim** [comment]
I deleted the overridden function "resolve" in WithScope.java because that was adding the view from the WITH clause into the scope of the main statement. This made the query pass by validation when the view was accessed from the main statement later without the necessary SELECT FROM [view].

File: core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java [hunk 1]
```diff
@@ -875,6 +877,20 @@ public final Sql sql(String sql) {
...
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-4092">[CALCITE-4092]
+   * NPE using WITH clause without a corresponding SELECT FROM</a>. */
+  @Test void testWithNotSelected() {
+    final String sql = "with emp2 as (select max(empno) as empno from emp where empno > 10)\n"
+        + "select * from emp where empno < emp2.empno";
+    try {
+      sql(sql).ok();
```
Comment (laurentgo) [inline core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java:887]: This test probably belongs more to `SqlValidatorTest` which has support to test for error (and validate error messages).`

Comment (James-Jeyun-Kim) [reply core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java]: You're right. I moved the test to SqlValidatorTest and made some adjustments accordingly.

**Pull request merged**